### PR TITLE
Fix cpu flash sdpa incorrect results when inputs' innermost dim is not contiguous

### PR DESCRIPTION
**Pull request by qqaatw**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #187506
* #185651

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### CODE REVIEW

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 1]
```diff
@@ -359,9 +359,9 @@ void cpu_flash_attention(
   //    -> (Batch x KV_seq_len   x KV_num_heads x Dim_per_head)
   // Value (Batch x KV_num_heads x KV_seq_len   x Dim_per_head)
   //    -> (Batch x KV_seq_len   x KV_num_heads x Dim_per_head)
-  at::Tensor query = q.transpose(1, 2);
-  at::Tensor key = k.transpose(1, 2);
-  at::Tensor value = v.transpose(1, 2);
+  const at::Tensor query = q.stride(-1) == 1 ? q.transpose(1, 2) : q.transpose(1, 2).contiguous();
```
Comment (Skylion007) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:362]: No way to fix this without a contiguous copy?

Comment (qqaatw) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:362]: I am not aware of a fix to work around it if we use cpublas, which requires a dimension to be contiguous. I could be wrong, if you know a clever way to fix it I would be happy to learn and adopt.

Here we also make `attn_mask` contiguous (the vectorized kernel handling mask has some trick to support stride == 0):

https://github.com/pytorch/pytorch/blob/de6a048169ec35cced2116a8d1527ed160598dc0/aten/src/ATen/native/cpu/FlashAttentionKernel.cpp#L226-L252

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -18223,8 +18223,6 @@ def sample_inputs_abs(op_info, device, dtype, requires_grad, op_kwargs=None, **k
         supports_cow_input_no_materialize_forward=False,
         decorators=[onlyCPU],
         skips=(
-            # Bug: noncontiguous inputs produce NaN outputs - implementation doesn't handle strides correctly
-            DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_noncontiguous_samples', device_type='cpu'),
         )
```
Comment (malfet) [inline torch/testing/_internal/common_methods_invocations.py]: Why not delete skips completely?

File: test/functorch/test_ops.py [hunk 3]
```diff
@@ -1516,6 +1514,9 @@ def test():
                     "index_fill"
                 ),  # aten::_unique hit the vmap fallback which is currently disabled
                 xfail("native_group_norm"),
+                xfail(
+                    "torch.ops.aten._scaled_dot_product_flash_attention_for_cpu"
+                ),  # aten::_scaled_dot_product_flash_attention_for_cpu hit the vmap fallback which is currently disabled
```
Comment (malfet) [inline test/functorch/test_ops.py:1524]: What's the point of moving this around?

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 4]
```diff
@@ -810,15 +810,20 @@ void cpu_flash_attention_backward(
   constexpr bool is_reduced_type = is_reduced_floating_point_v<scalar_t>;
   using accum_t = at::opmath_type<scalar_t>;
   using Vec = vec::Vectorized<accum_t>;
+
+  // Query (Batch x Q_seq_len  x Num_heads    x Dim_per_head)
+  // Key   (Batch x KV_seq_len x KV_num_heads x Dim_per_head)
+  // Value (Batch x KV_seq_len x KV_num_heads x Dim_per_head)
+  const at::Tensor query = q.stride(-1) == 1 ? q : q.contiguous();
```
Comment (malfet) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:817]: Err? Why just last stride? Do you have a test that checks it works for if other dims are sparse but last one is contiguous?

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp (hunk 1 above)
Comment (malfet) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:362]: I'm again, confused by this condition. Why `.stride(-1)` check? And how about broadcasts? Is it also prohibited?

File: test/functorch/test_ops.py (hunk 3 above)
Comment (qqaatw) [reply test/functorch/test_ops.py:1524]: They are different tests.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (qqaatw) [reply torch/testing/_internal/common_methods_invocations.py]: oops, will fix it

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp (hunk 4 above)
Comment (qqaatw) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:817]: Because `gemm` in blas only cares about the 2-D matrix shape (last 2 dimensions) to compute on, the contiguity of the last dimension to perform efficient dot product, and the leading dimension (LD), which dictates how many strides to skip to advance to the next row / column (depending on whether it's row or  column major).

So in our case here, let's say we have a contiguous tensor

shape: `(2, 2, 2, 2)` stride: `(8, 4, 2, 1)`

even though we have the dimension 1 and 2 permuted, 

shape: `(2, 2, 2, 2)` stride: `(8, 2, 4, 1)`

as long as the `gemm` knows the matrix shape `(2, 2)`, the leading dimension `4` (how many strides to advance to the next row/column), and the pointer to the matrix to start with, it doesn't really care about whether the entire tensor is contiguous or not.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp (hunk 1 above)
Comment (qqaatw) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:362]: Explained `.stride(-1)` in another comment for gemm. For attn_mask, it's likely because the kernel is vectorized so the last dim contiguity is required.

About broadcasting, the broadcasting condition is constrained below, and the `expand` is used to broadcast last two dimensions. `attn_mask` is later applied to `qk` row-by-row, handling the broadcasting for the first two dimensions.

https://github.com/pytorch/pytorch/blob/1aa05b63f5f8d280b4177db3ed913aa56b62c4a1/aten/src/ATen/native/transformers/sdp_utils_cpp.h#L269-L320

**Review by zou3519 — COMMENTED** [review]
test_ops changes lgtm, deferring to everyone else for the actual review

**Comment by qqaatw** [comment]
@claude review

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp (hunk 4 above)
Comment (malfet) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:817]: There is a storage_contiguous() or `dense_in_storage` or something, which you should use here. What I'm afraid of is shape (2, 2, 2, 2), but say stride `(0, 0, 8, 1)` or `(64, 8, 4, 1)` and you'll still treat this one as contiguous

**Comment by qqaatw** [comment]
@claude analyze @malfet's assumption and validate the implementation.

**Comment by qqaatw** [comment]
@claude validate the implementation for this stride: (4, 2, 0, 1), including the system blas path & pytorch's cpu blas path.

**Review by malfet — APPROVED** [review]
Sure, why not, though again, it still feels like a some sort of an insane logic that should be handled elsewhere. I.e. if it cause to `torch.mm`, this operator already can handle all strides and sizes of input matrices
